app/assets/src/components/views/discovery/mapping/MapPlayground.jsx

```diff
@@ -1,78 +1,153 @@
+import { Marker, Popup as MapPopup } from "react-map-gl";
+import { get, remove } from "lodash/fp";
 
 import BaseMap from "~/components/views/discovery/mapping/BaseMap";
 import CircleMarker from "~/components/views/discovery/mapping/CircleMarker";
 
+import cs from "./map_playground.scss";
+
 class MapPlayground extends React.Component {
   constructor(props) {
     super(props);
+    const { results } = this.props;
 
     // Load demo data as locations->samples
-    const { results } = this.props;
-    const locationsToSamples = {};
+    const locationsToItems = {};
     results.forEach(result => {
       // Match locations that look like coordinates separated by a comma
-      const loc = result.location.replace(/_/g, ", ");
-      if (loc.match(/^[-0-9.]+,(\s)?[-0-9.]+?$/)) {
-        const formatted = {
+      let locationName = result.location.replace(/_/g, ", ");
+      const match = /^([-0-9.]+),\s?([-0-9.]+)$/g.exec(locationName);
+      if (match) {
+        let [lat, lon] = this.parseLatLon(match[1], match[2]);
+        if (!(lat && lon)) return;
+        locationName = `${lat}, ${lon}`;
+        const item = {
           name: result.name,
           id: result.id
         };
-        if (locationsToSamples.hasOwnProperty(loc)) {
-          locationsToSamples[loc].push(formatted);
+
+        // Group items under the location name
+        if (locationsToItems.hasOwnProperty(locationName)) {
+          locationsToItems[locationName].items.push(item);
         } else {
-          locationsToSamples[loc] = [formatted];
+          locationsToItems[locationName] = {
+            lat: lat,
+            lon: lon,
+            items: [item]
+          };
         }
       }
     });
 
     this.state = {
-      locationsToSamples,
-      viewport: {}
+      locationsToItems: locationsToItems,
+      viewport: {},
+      popups: []
     };
   }
 
+  parseLatLon = (lat, lon) => {
+    // Round the coordinates for some minimal aggregation
+    lat = parseFloat(parseFloat(lat).toFixed(2));
+    lon = parseFloat(parseFloat(lon).toFixed(2));
+    // Reject invalid coordinates
+    if (lat < -90 || lat > 90 || lon < -180 || lon > 180) {
+      return [null, null];
+    } else {
+      return [lat, lon];
+    }
+  };
+
   updateViewport = viewport => {
     this.setState({ viewport });
   };
 
-  renderMarker = (markerData, index) => {
+  renderMarker = (marker, index) => {
     const { viewport } = this.state;
-    const [lat, lon] = markerData[0].split(",");
-    const pointCount = markerData[1].length;
-    const minSize = 14;
+    const [locationName, markerData] = marker;
+    const pointCount = markerData.items.length;
+    const minSize = 12;
     // Scale based on the zoom and point count (zoomed-in = higher zoom)
     const markerSize = Math.max(
-      pointCount * (get("zoom", viewport) || 3),
+      Math.log(pointCount) / Math.log(1.5) * (get("zoom", viewport) || 3),
       minSize
     );
+    const hoverContent = (
+      <div>
```

Comment:
Could you use <DataTooltip> here> To try to make the styling more consistent.

Reply:
👍 changed. Reworked to put the hover management in state and keep everything inside the MapboxGL map instead of having a popup outside